File: src/preprocess.rs
//! Image preprocessing: load, resize to model input size.
//!
//! Uses plain bilinear sampling (matching OpenCV `INTER_LINEAR`) rather than
//! `image::imageops::resize`, whose Triangle filter applies anti-aliasing on
//! downscale and produces different pixels than the Python pipeline.

use image::GenericImageView;

use crate::error::Result;

/// Plain bilinear resize of an RGB8 buffer — bit-compatible with
/// `cv2.resize(..., interpolation=cv2.INTER_LINEAR)`.
fn resize_bilinear_rgb(
    src: &[u8],
    src_w: usize,
    src_h: usize,
    dst_w: usize,
    dst_h: usize,
) -> Vec<u8> {
    let mut dst = vec![0u8; dst_w * dst_h * 3];
    let sx = src_w as f64 / dst_w as f64;
    let sy = src_h as f64 / dst_h as f64;

    for y in 0..dst_h {
        let fy = ((y as f64 + 0.5) * sy - 0.5).max(0.0);
        let y0 = (fy.floor() as usize).min(src_h - 1);
        let y1 = (y0 + 1).min(src_h - 1);
        let wy = fy - y0 as f64;

        for x in 0..dst_w {
            let fx = ((x as f64 + 0.5) * sx - 0.5).max(0.0);
            let x0 = (fx.floor() as usize).min(src_w - 1);
            let x1 = (x0 + 1).min(src_w - 1);
            let wx = fx - x0 as f64;

            let p00 = (y0 * src_w + x0) * 3;
            let p01 = (y0 * src_w + x1) * 3;
            let p10 = (y1 * src_w + x0) * 3;
            let p11 = (y1 * src_w + x1) * 3;
            let d = (y * dst_w + x) * 3;

            for c in 0..3 {
                let v = src[p00 + c] as f64 * (1.0 - wx) * (1.0 - wy)
                    + src[p01 + c] as f64 * wx * (1.0 - wy)
                    + src[p10 + c] as f64 * (1.0 - wx) * wy
                    + src[p11 + c] as f64 * wx * wy;
                dst[d + c] = (v + 0.5) as u8;
            }
        }
    }
    dst
}

/// Load an image and resize it to the model's input dimensions.
///
/// Returns a flat RGB buffer (NHWC layout) suitable for RKNN input.
pub fn preprocess(img_path: &str, target_w: usize, target_h: usize) -> Result<Vec<u8>> {
    let img = image::open(img_path)?;
    let (orig_w, orig_h) = img.dimensions();
    log::info!(
        "Preprocess: {}x{} -> {}x{} (bilinear)",
        orig_w,
        orig_h,
        target_w,
        target_h
    );

    let rgb = img.to_rgb8();
    Ok(resize_bilinear_rgb(
        rgb.as_raw(),
        orig_w as usize,
        orig_h as usize,
        target_w,
        target_h,
    ))
}

/// Load an image and return it as a DynamicImage (for point cloud color lookup).
pub fn load_image(img_path: &str) -> Result<image::DynamicImage> {
    Ok(image::open(img_path)?)
}

File: cpp/src/preprocess.cpp
#include "preprocess.hpp"

#include <algorithm>
#include <cmath>
#include <iostream>

#include <opencv2/imgproc.hpp>
#include <opencv2/imgcodecs.hpp>

static constexpr int STRIDE = 32;

static int align32(float x) {
    return std::max(static_cast<int>(std::round(x / STRIDE) * STRIDE), 32);
}

void computeRectInput(int src_h, int src_w, int model_max_dim,
                      int &out_h, int &out_w) {
    float scale = static_cast<float>(model_max_dim) / std::max(src_h, src_w);
    out_h = align32(src_h * scale);
    out_w = align32(src_w * scale);
}


cv::Mat preprocess(const cv::Mat &bgr, int target_w, int target_h) {
    int src_h = bgr.rows;
    int src_w = bgr.cols;
    int model_max_dim = std::max(target_h, target_w);

    int rect_h, rect_w;
    computeRectInput(src_h, src_w, model_max_dim, rect_h, rect_w);

    cv::Mat resized;
    if (rect_h == target_h && rect_w == target_w) {
        cv::resize(bgr, resized, cv::Size(target_w, target_h), 0, 0, cv::INTER_LINEAR);
    } else {
        std::cerr << "Warning: Image aspect ratio (" << src_w << ":" << src_h
                  << ") does not match model aspect ratio (" << target_w << ":" << target_h
                  << "). Rect input would be " << rect_w << "x" << rect_h
                  << " but model expects " << target_w << "x" << target_h << "."
                  << std::endl;
        cv::resize(bgr, resized, cv::Size(target_w, target_h), 0, 0, cv::INTER_LINEAR);
    }

    // BGR -> RGB, keep NHWC layout
    cv::Mat rgb;
    cv::cvtColor(resized, rgb, cv::COLOR_BGR2RGB);

    std::cout << "Preprocess: " << src_w << "x" << src_h << " -> "
              << rgb.cols << "x" << rgb.rows << " (bilinear)" << std::endl;

    return rgb; // uint8, H x W x 3, NHWC
}

cv::Mat preprocess(const std::string &img_path, int target_w, int target_h) {
    cv::Mat bgr = cv::imread(img_path, cv::IMREAD_COLOR);
    if (bgr.empty()) {
        throw std::runtime_error("Failed to load image: " + img_path);
    }
    return preprocess(bgr, target_w, target_h);
}

cv::Mat loadImage(const std::string &img_path) {
    cv::Mat img = cv::imread(img_path, cv::IMREAD_COLOR);
    if (img.empty()) {
        throw std::runtime_error("Failed to load image: " + img_path);
    }
    return img;
}
